nanofi/src/core/cstream.c

```diff
@@ -0,0 +1,203 @@
+}
+
+int write_buffer(const uint8_t *value, int len, cstream * stream) {
+  int ret = 0, bytes = 0;
+
+  while (bytes < len) {
+    ret = send(stream->socket_, value + bytes, len - bytes, 0);
+    // check for errors
+    if (ret <= 0) {
+      logc(err, "Could not send to %d, error: %s", stream->socket_, strerror(errno));
+      close_stream(stream);
+      return ret;
+    }
+    bytes += ret;
+  }
+
+  if (ret)
+    logc(trace, "Send data size %d over socket %d", len, stream->socket_);
+  return bytes;
+}
+
+int read_buffer(uint8_t *buf, int len, cstream * stream) {
+  int32_t total_read = 0;
+  while (len) {
+    int bytes_read = recv(stream->socket_, buf, len, 0);
+    logc(trace, "Recv call %d", bytes_read);
+    if (bytes_read <= 0) {
+      if (bytes_read == 0) {
+        logc(debug, "Other side hung up on %d",stream->socket_);
+      } else {
+        if (errno == EAGAIN || errno == EWOULDBLOCK) {
+          // continue
+          return -2;
+        }
+        logc(err, "Could not recv on %d, error: %s", stream->socket_, strerror(errno));
+      }
+      return -1;
+    }
+    len -= bytes_read;
+    buf += bytes_read;
+    total_read += bytes_read;
+  }
+  return total_read;
+}
+
+int writeUTF(const char * cstr, uint64_t len, enum Bool widen, cstream * stream) {
+  if (len > 65535) {
+    return -1;
+  }
+
+  int ret;
+  if (!widen) {
+    uint16_t shortlen = len;
+    ret = write_uint16_t(shortlen, stream);
+  } else {
+    ret = write_uint32_t(len, stream);
+  }
+
+  if(len == 0 || ret < 0) {
+    return ret;
+  }
+
+  const uint8_t *underlyingPtr = (const uint8_t *)cstr;
+
+  if (!widen) {
+    uint16_t short_length = len;
+    ret = write_buffer(underlyingPtr, short_length, stream);
+  } else {
+    ret = write_buffer(underlyingPtr, len, stream);
+  }
+  return ret;
+}
+
+int read_uint8_t(uint8_t *value, cstream * stream) {
+  uint8_t val;
+  int ret = read_buffer(&val, sizeof(uint8_t), stream);
+  if(ret == sizeof(uint8_t)) {
+    *value = val;
+  }
+  return ret;
+}
+int read_uint16_t(uint16_t *value, cstream * stream) {
+  uint16_t val;
+  int ret = read_buffer((uint8_t*)&val, sizeof(uint16_t), stream);
+  if(ret == sizeof(uint16_t)) {
+    *value = ntohs(val);
+  }
+  return ret;
+}
+int read_uint32_t(uint32_t *value, cstream * stream) {
+  uint32_t val;
+  int ret = read_buffer((uint8_t*)&val, sizeof(uint32_t), stream);
+  if(ret == sizeof(uint32_t)) {
+    *value = ntohl(val);
+  }
+  return ret;
+}
+
+int readUTFLen(uint32_t * utflen, cstream * stream) {
+  int ret = 1;
+  uint16_t shortLength = 0;
+  ret = read_uint16_t(&shortLength, stream);
+  if (ret > 0) {
+    *utflen = shortLength;
+  }
+  return ret;
+}
+
+int readUTF(char * buf, uint64_t buflen, cstream * stream) {
+  //return stream->impl->readData((uint8_t*)buf, buflen);
+  return read_buffer((uint8_t*)buf, buflen, stream);
+}
+
+void close_stream(cstream * stream) {
+  if(stream != NULL && stream->socket_ != -1) {
+    shutdown(stream->socket_, SHUT_RDWR);
```

Comment:
I think this leaks file descriptors. Shutdown is not enough in itself, resources are freed by close().

Reply:
Good shout, thanks, fixed!